Create v2 of node distribution standard (issues/#494) (#524)
SovereignCloudStack/standards: Standards/scs-0214-v1-k8s-node-distribution.md

  can also be scaled vertically first before scaling horizontally.
- Worker node distribution MUST be indicated to the user through some kind of labeling
  in order to enable (anti)-affinity for workloads over "failure zones".
- To provide data about this node distribution and be able to test and verify compliance with this standard,
  providers MUST label their K8s nodes with the labels "topology.kubernetes.io/region",

>   "topology.kubernetes.io/zone" and "topology.scs.openstack.org/host-id"

Reviewer: [Here](https://github.com/SovereignCloudStack/issues/issues/540#issuecomment-1997330524), @martinmo proposed changing the label name to `topology.scs.community/host-id`

As I do not have a strong opinion here, I just changed this name in the related e2e test https://github.com/SovereignCloudStack/k8s-cluster-api-provider/commit/ffb722b42e99977d6cb42817993e11558d816012

So use the same name or raise your voice in the related issue if you do not agree with the latest proposal.

Reviewer 2: I agree. I think so does @cah-hbaum, but I don't know for sure.

Author: Also agree, I only had this information later on. I'm gonna change it with the next commit.

  in order to provide the expected data on the KaaS layer.
  "topology.kubernetes.io/region" and "topology.kubernetes.io/zone" correspond with the labels
  described in the [K8s labels documentation][k8s-labels-docs] and provide logical zones of failure
  as well as a domain made up of one or more zones. How these are defined is up to the provider and
  the fields are most of the time autopopulated by either the kubelet, the cloud controller or another
  external mechanism.

>   "topology.scs.openstack.org/host-id" is an SCS-specific label, which MUST contain the hostID of the

Reviewer: adjust the label name based on the comment above

>   physical machine running the hypervisor and not the hostID of a virtual machine.

Reviewer 2: Three things:

- I would prefer more structure (use a list for the labels)
- then state for each label what the value should be (even though it's not authoritative for region and zone)
- maybe don't narrow it down to verifying compliance so much; it's also desirable in its own right for maximum transparency (and it seems t be quite common already, maybe for this very reason)

Reviewer 2: Oh sorry, I didn't mean concrete values, just an explanation what kind of information we expect, namely the zone and the region -- you know, this may sound redundant, but I want to be extra clear that the label is what the label is :)

Author: Ah ok got it

    like the cloud controller.
  
  - "topology.kubernetes.io/region"  
    Corresponds with the label described in [K8s labels documentation][k8s-labels-docs].
    It describes the combination of one or more failure zones into a region or domain, therefore
    showing a larger entity of logical failure zone. An example for this could be a building
    containing racks that are put into such a zone, since they're all prone to failure, if e.g.
    the power for the building is cut. How this is defined exactly is also up to the provider.
    The field gets autopopulated most of the time by either the kubelet or external mechanisms
    like the cloud controller.
  
  - "topology.scs.community/host-id"  
    This is an SCS-specific label, which MUST contain the hostID of the physical machine running
    the hypervisor and not the hostID of a virtual machine. This helps identify the distribution
    over underlying physical machines, which would be masked if VM hostIDs would be used.

## Conformance Tests

The script `k8s-node-distribution-check.py` checks the nodes available with a user-provided

> kubeconfig file. It then determines based on the labels `topology.scs.openstack.org/host-id`,

Reviewer: ```suggestion
kubeconfig file. It then determines based on the labels `topology.scs.community/host-id,
```

Author: Oh yeah I forgot that. Fixed it now.

    in the same electrical circuit or multiple machines bound to the internet through a
    singular network structure. How this is defined exactly is up to the plans of the provider.
    The field gets autopopulated most of the time by either the kubelet or external mechanisms
    like the cloud controller.

  - "topology.kubernetes.io/region"

    Corresponds with the label described in [K8s labels documentation][k8s-labels-docs].
    It describes the combination of one or more failure zones into a region or domain, therefore
    showing a larger entity of logical failure zone. An example for this could be a building
    containing racks that are put into such a zone, since they're all prone to failure, if e.g.
    the power for the building is cut. How this is defined exactly is also up to the provider.
    The field gets autopopulated most of the time by either the kubelet or external mechanisms
    like the cloud controller.

  - "topology.scs.community/host-id"

    This is an SCS-specific label, which MUST contain the hostID of the physical machine running
    the hypervisor and not the hostID of a virtual machine. This helps identify the distribution

>     over underlying physical machines, which would be masked if VM hostIDs would be used.

Reviewer 2: I like the new version very much! One minor thing with this paragraph: we have to say what we mean by `hostID`. We should say that this is more or less an arbitrary identifier for the host (need not be identical with the hostname or anything tangible), but it must be unique to the host so it is clear whether two nodes are placed on the same host.

Author: ```

## Conformance Tests

The script `k8s-node-distribution-check.py` checks the nodes available with a user-provided

> kubeconfig file. It then determines based on the labels `topology.scs.community/host-id`,
> `topology.kubernetes.io/zone`, `topology.kubernetes.io/region` and `node-role.kubernetes.io/control-plane`,
> if a distribution of the available nodes is present. If this isn't the case, the script produces an error.

Reviewer 2: ```suggestion
kubeconfig file. Based on the labels `topology.scs.community/host-id`,
`topology.kubernetes.io/zone`, `topology.kubernetes.io/region` and `node-role.kubernetes.io/control-plane`,
the script then determines whether the nodes are distributed according to this standard.
If this isn't the case, the script produces an error.
```

    in the same electrical circuit or multiple machines bound to the internet through a
    singular network structure. How this is defined exactly is up to the plans of the provider.
    The field gets autopopulated most of the time by either the kubelet or external mechanisms
    like the cloud controller.

  - "topology.kubernetes.io/region"

    Corresponds with the label described in [K8s labels documentation][k8s-labels-docs].
    It describes the combination of one or more failure zones into a region or domain, therefore
    showing a larger entity of logical failure zone. An example for this could be a building
    containing racks that are put into such a zone, since they're all prone to failure, if e.g.
    the power for the building is cut. How this is defined exactly is also up to the provider.
    The field gets autopopulated most of the time by either the kubelet or external mechanisms
    like the cloud controller.

  - "topology.scs.community/host-id"

    This is an SCS-specific label, which MUST contain the hostID of the physical machine running
    the hypervisor and not the hostID of a virtual machine. The hostID is an arbitrary identifier,

>     which doesn't need to contain things like hostname, but it should nonetheless be unique to the host.

Reviewer 2: ```suggestion
    which need not contain the actual hostname, but it should nonetheless be unique to the host.
```

Author: ```
Here, the hostID is an arbitrary identifier, which doesn't need to contain the actual hostname, but it should nonetheless be unique to the host.
```
Is this version ok @mbuechse ?

Reviewer 2: What's wrong with `need not contain`?

Reviewer 2: https://english.stackexchange.com/questions/29409/why-use-need-not-instead-of-do-not-need-to

Author: I guess mainly personal preference, auxiliary verb usage doesn't sound that good to me.
But no hard preference on that.

Author: Updated it.
